File: smash/fcore/derived_type/mwd_setup.f90
!%      (MWD) Module Wrapped and Differentiated.
!%
!%      Type
!%      ----
!%
!%      - SetupDT
!%          All user setup informations. See default values in _constant.py DEFAULT_SETUP
!%
!%          ========================== =====================================
!%          `Variables`                Description
!%          ========================== =====================================
!%          ``snow_module``            Snow module
!%          ``hydrological_module``    Hydrological module
!%          ``routing_module``         Routing module
!%          ``serr_mu_mapping``        Mapping for structural error model
!%          ``serr_sigma_mapping``     Mapping for structural error model
!%          ``dt``                     Solver time step        [s]
!%          ``start_time``             Simulation start time   [%Y%m%d%H%M]
!%          ``end_time``               Simulation end time     [%Y%m%d%H%M]
!%          ``adjust_interception``    Adjust interception reservoir capacity
!%          ``compute_mean_atmos``     Compute mean atmospheric data for each gauge
!%          ``read_qobs``              Read observed discharge
!%          ``qobs_directory``         Observed discharge directory path
!%          ``read_prcp``              Read precipitation
!%          ``prcp_format``            Precipitation format
!%          ``prcp_conversion_factor`` Precipitation conversion factor
!%          ``prcp_directory``         Precipiation directory path
!%          ``prcp_access``            Precipiation access tree
!%          ``read_pet``               Read potential evapotranspiration
!%          ``pet_format``             Potential evapotranspiration format
!%          ``pet_conversion_factor``  Potential evapotranpisration conversion factor
!%          ``pet_directory``          Potential evapotranspiration directory path
!%          ``pet_access``             Potential evapotranspiration access tree
!%          ``daily_interannual_pet``  Read daily interannual potential evapotranspiration
!%          ``read_snow``              Read snow
!%          ``snow_format``            Snow format
!%          ``snow_conversion_factor`` Snow conversion factor
!%          ``snow_directory``         Snow directory path
!%          ``snow_access``            Snow access tree
!%          ``read_temp``              Read temperatur
!%          ``temp_format``            Temperature format
!%          ``temp_directory``         Temperature directory path
!%          ``temp_access``            Temperature access tree
!%          ``prcp_partitioning``      Precipitation partitioning
!%          ``sparse_storage``         Forcing sparse storage
!%          ``read_descriptor``        Read descriptor map(s)
!%          ``descriptor_format``      Descriptor maps format
!%          ``descriptor_directory``   Descriptor maps directory
!%          ``descriptor_name``        Descriptor maps names
!%          ``structure``              Structure combaining all modules
!%          ``snow_module_present``    Presence of snow module
!%          ``ntime_step``             Number of time steps
!%          ``nd``                     Number of descriptor maps
!%          ``nrrp``                   Number of rainfall-runoff parameters
!%          ``nrrs``                   Number of rainfall-runoff states
!%          ``nsep_mu``                Number of structural error parameters for mu
!%          ``nsep_sigma``             Number of structural error parameters for sigma
!%          ``nqz``                    Size of the temporal buffer for discharge grids
!%
!%      Subroutine
!%      ----------
!%
!%      - SetupDT_initialise
!%      - SetupDT_copy

module mwd_setup

    use md_constant !% only: sp, lchar

    implicit none

    type SetupDT

        !% Notes
        !% -----
        !% SetupDT Derived Type.

        ! User variables
        character(lchar) :: snow_module = "..." !$F90W char
        character(lchar) :: hydrological_module = "..." !$F90W char
        character(lchar) :: routing_module = "..."!$F90W char

        character(lchar) :: serr_mu_mapping = "..." !$F90W char
        character(lchar) :: serr_sigma_mapping = "..." !$F90W char

        real(sp) :: dt = -99._sp

        character(lchar) :: start_time = "..." !$F90W char
        character(lchar) :: end_time = "..." !$F90W char

        logical :: adjust_interception = .true.
        logical :: compute_mean_atmos = .true.

        logical :: read_qobs = .false.
        character(2*lchar) :: qobs_directory = "..." !$F90W char

        logical :: read_prcp = .false.
        character(lchar) :: prcp_format = "..." !$F90W char
        real(sp) :: prcp_conversion_factor = 1._sp
        character(2*lchar) :: prcp_directory = "..." !$F90W char
        character(lchar) :: prcp_access = "..." !$F90W char

        logical :: read_pet = .false.
        character(lchar) :: pet_format = "..." !$F90W char
        real(sp) :: pet_conversion_factor = 1._sp
        character(2*lchar) :: pet_directory = "..." !$F90W char
        character(lchar) :: pet_access = "..." !$F90W char
        logical :: daily_interannual_pet = .false.

        logical :: read_snow = .false.
        character(lchar) :: snow_format = "..." !$F90W char
        real(sp) :: snow_conversion_factor = 1._sp
        character(2*lchar) :: snow_directory = "..." !$F90W char
        character(lchar) :: snow_access = "..." !$F90W char

        logical :: read_temp = .false.
        character(lchar) :: temp_format = "..." !$F90W char
        character(2*lchar) :: temp_directory = "..." !$F90W char
        character(lchar) :: temp_access = "..." !$F90W char

        logical :: prcp_partitioning = .false.

        logical :: sparse_storage = .false.

        logical :: read_descriptor = .false.
        character(lchar) :: descriptor_format = "..." !$F90W char
        character(2*lchar) :: descriptor_directory = "..." !$F90W char
        character(lchar), allocatable, dimension(:) :: descriptor_name !$F90W char-array

        ! Post processed variables
        character(lchar) :: structure = "..." !$F90W char
        logical :: snow_module_present = .false.

        integer :: ntime_step = -99
        integer :: nd = -99
        integer :: nrrp = -99
        integer :: nrrs = -99
        integer :: nsep_mu = -99
        integer :: nsep_sigma = -99
        integer :: nqz = -99

    end type SetupDT

contains

    subroutine SetupDT_initialise(this, nd)

        !% Notes
        !% -----
        !% SetupDT initialisation subroutine

        implicit none

        type(SetupDT), intent(inout) :: this
        integer, intent(in) :: nd

        this%nd = nd

        allocate (this%descriptor_name(this%nd))
        this%descriptor_name = "..."

    end subroutine SetupDT_initialise

    subroutine SetupDT_copy(this, this_copy)

        implicit none

        type(SetupDT), intent(in) :: this
        type(SetupDT), intent(out) :: this_copy

        this_copy = this

    end subroutine SetupDT_copy

end module mwd_setup

File: Pytorch_playground/pycore/derived_type/mwd_setup.py
class SetupDT:
    def __init__(self, nd):
        # Define attributes
        self.snow_module = "..."  # Snow module
        self.hydrological_module = "..."  # Hydrological module
        self.routing_module = "..."  # Routing module
        self.serr_mu_mapping = "..."  # Mapping for structural error model
        self.serr_sigma_mapping = "..."  # Mapping for structural error model
        self.dt = -99.0  # Solver time step [s]
        self.start_time = "..."  # Simulation start time [%Y%m%d%H%M]
        self.end_time = "..."  # Simulation end time [%Y%m%d%H%M]
        self.adjust_interception = True  # Adjust interception reservoir capacity
        self.compute_mean_atmos = True  # Compute mean atmospheric data for each gauge
        self.read_qobs = False  # Read observed discharge
        self.qobs_directory = "..."  # Observed discharge directory path
        self.read_prcp = False  # Read precipitation
        self.prcp_format = "..."  # Precipitation format
        self.prcp_conversion_factor = 1.0  # Precipitation conversion factor
        self.prcp_directory = "..."  # Precipiation directory path
        self.prcp_access = "..."  # Precipiation access tree
        self.read_pet = False  # Read potential evapotranspiration
        self.pet_format = "..."  # Potential evapotranspiration format
        self.pet_conversion_factor = 1.0  # Potential evapotranpisration conversion factor
        self.pet_directory = "..."  # Potential evapotranspiration directory path
        self.pet_access = "..."  # Potential evapotranspiration access tree
        self.daily_interannual_pet = False  # Read daily interannual potential evapotranspiration
        self.read_snow = False  # Read snow
        self.snow_format = "..."  # Snow format
        self.snow_conversion_factor = 1.0  # Snow conversion factor
        self.snow_directory = "..."  # Snow directory path
        self.snow_access = "..."  # Snow access tree
        self.read_temp = False  # Read temperatur
        self.temp_format = "..."  # Temperature format
        self.temp_directory = "..."  # Temperature directory path
        self.temp_access = "..."  # Temperature access tree
        self.read_sm = False  # Read soil moisture (default: .false.)
        self.sm_format = "..."  # Soil moisture format (default: 'tif')
        self.sm_conversion_factor = 1.0  # Soil moisture format conversion factor (default: 1)
        self.sm_directory = "..."  # Soil moisture directory path (default: '...')
        self.sm_access = "..."  # Soil moisture access tree
        self.sm_metric = "..."  # Soil moisture metric
        self.prcp_partitioning = False  # Precipitation partitioning
        self.sparse_storage = False
        self.read_descriptor = False  # Read descriptor map(s)
        self.descriptor_format = "..."  # Descriptor maps format
        self.descriptor_directory = "..."  # Descriptor maps directory
        self.descriptor_name = ["..."] * nd  # Initialize descriptor_name with nd elements
        self.structure = "..."  # Structure combining all modules
        self.snow_module_present = False  # Presence of snow module
        self.ntime_step = -99  # Number of time steps
        self.nd = nd  # Number of descriptor maps
        self.nrrp = -99  # Number of rainfall-runoff parameters
        self.nrrs = -99  # Number of rainfall-runoff states
        self.nsep_mu = -99  # Number of structural error parameters for mu
        self.nsep_sigma = -99  # Number of structural error parameters for sigma
        self.nqz = -99  # Size of the temporal buffer for discharge grids
        self.maxiter = -99

    def copy(self):
        return SetupDT(self.nd)
